File: backend-ci/gestion/services/imam_service.py
from django.db import transaction
from django.shortcuts import get_object_or_404
from gestion.models.imam import Imam


class ImamService:

    @staticmethod
    @transaction.atomic
    def create(data):
        """Créer un nouvel imam"""
        imam = Imam.objects.create(
            nom_complet=data['nom_complet'],
            biographie=data.get('biographie', ''),
            photo_url=data.get('photo_url', ''),
            periode=data.get('periode', '')
        )
        return imam

    @staticmethod
    def get_all():
        """Récupérer tous les imams"""
        return Imam.objects.all()

    @staticmethod
    def get_by_id(imam_id):
        """Récupérer un imam par son ID"""
        return get_object_or_404(Imam, id=imam_id)

    @staticmethod
    def update(imam, data):
        """Mettre à jour un imam"""
        for key, value in data.items():
            setattr(imam, key, value)
        imam.save()
        return imam

    @staticmethod
    def delete(imam):
        """Supprimer un imam"""
        imam.delete()

File: frontend-ci/src/services/imam.service.ts
/**
 * Service pour les imams
 * Correspond à gestion/services/imam_service.py
 */

import apiClient from './api';
import { Imam, SaveImamDTO, UpdateImamDTO, CreateResponse } from '../types';

const ENDPOINTS = {
  BASE: '/imams/',
  DETAIL: (id: number) => `/imams/${id}/`,
};

export const imamService = {
  /**
   * Récupérer tous les imams
   */
  async getAll(): Promise<Imam[]> {
    const response = await apiClient.get<Imam[]>(ENDPOINTS.BASE);
    return response.data;
  },

  /**
   * Récupérer un imam par son ID
   */
  async getById(id: number): Promise<Imam> {
    const response = await apiClient.get<Imam>(ENDPOINTS.DETAIL(id));
    return response.data;
  },

  /**
   * Créer un nouvel imam
   */
  async create(data: SaveImamDTO): Promise<CreateResponse> {
    const response = await apiClient.post<CreateResponse>(ENDPOINTS.BASE, data);
    return response.data;
  },

  /**
   * Mettre à jour un imam
   */
  async update(id: number, data: UpdateImamDTO): Promise<Imam> {
    const response = await apiClient.put<Imam>(ENDPOINTS.DETAIL(id), data);
    return response.data;
  },

  /**
   * Supprimer un imam
   */
  async delete(id: number): Promise<void> {
    await apiClient.delete(ENDPOINTS.DETAIL(id));
  },
};

export default imamService;
